### ISSUE TITLE
xdg-desktop-portal and cage don't play well together

### ISSUE THREAD

**Issue by pescepalla** [flatpak/xdg-desktop-portal#1115]
I use greetd-regreet with cage as a Wayland login manager.

It takes up to 30 seconds before anything other than a black screen shows up. 

The issue does not occur if I uninstall xdg-desktop-portal or use Sway as a greetd-regreet backend, so I'm not sure whether this is a problem with cage or xdg-desktop-portal.

OS: Arch Linux
WM: Hyprland
GPU: NVIDIA 3070

**Comment by smcv** [comment]
Please check the systemd journal for any relevant warning or error messages? There's very little we can do with "it's slow".

**Comment by pescepalla** [comment]
I have reinstalled `xdg-desktop-portal `and `xdg-desktop-portal-wlr` for testing reasons.

I hope the following is useful:
```
$ journalctl | grep xdg-desktop-portal
... (I've cut everything out except the latest reboot) ...
Sep 27 13:45:00 stellaris dbus-daemon[884]: [session uid=971 pid=884] Activating via systemd: service name='org.freedesktop.portal.Desktop' unit='xdg-desktop-portal.service' requested by ':1.0' (uid=971 pid=878 comm="regreet")
Sep 27 13:45:00 stellaris dbus-daemon[884]: [session uid=971 pid=884] Activating via systemd: service name='org.freedesktop.portal.Documents' unit='xdg-document-portal.service' requested by ':1.2' (uid=971 pid=887 comm="/usr/lib/xdg-desktop-portal")
Sep 27 13:45:00 stellaris dbus-daemon[575]: [system] Activating via systemd: service name='org.freedesktop.RealtimeKit1' unit='rtkit-daemon.service' requested by ':1.17' (uid=971 pid=887 comm="/usr/lib/xdg-desktop-portal")
Sep 27 13:45:00 stellaris dbus-daemon[884]: [session uid=971 pid=884] Activating via systemd: service name='org.freedesktop.impl.portal.desktop.wlr' unit='xdg-desktop-portal-wlr.service' requested by ':1.2' (uid=971 pid=887 comm="/usr/lib/xdg-desktop-portal")
**Sep 27 13:45:25 stellaris dbus-daemon[884]: [session uid=971 pid=884] Activating service name='org.freedesktop.secrets' requested by ':1.2' (uid=971 pid=887 comm="/usr/lib/xdg-desktop-portal")**
Sep 27 13:45:39 stellaris dbus-daemon[1057]: [session uid=1000 pid=1057] Activating via systemd: service name='org.freedesktop.portal.Desktop' unit='xdg-desktop-portal.service' requested by ':1.13' (uid=1000 pid=1206 comm="thunar --daemon")
Sep 27 13:45:39 stellaris dbus-daemon[1057]: [session uid=1000 pid=1057] Activating via systemd: service name='org.freedesktop.portal.Documents' unit='xdg-document-portal.service' requested by ':1.20' (uid=1000 pid=1311 comm="/usr/lib/xdg-desktop-portal")
Sep 27 13:45:39 stellaris dbus-daemon[1057]: [session uid=1000 pid=1057] Activating via systemd: service name='org.freedesktop.impl.portal.desktop.wlr' unit='xdg-desktop-portal-wlr.service' requested by ':1.20' (uid=1000 pid=1311 comm="/usr/lib/xdg-desktop-portal")
```
It looks like the system takes 25 seconds to activate `org.freedesktop.secrets`

**Comment by smcv** [comment]
Please check for messages from any other components around that time, not just xdg-desktop-portal. The D-Bus message bus (usually dbus-daemon or dbus-broker), systemd, whatever is your implementation of org.freedesktop.secrets (probably gnome-keyring or KWallet?), and whatever is your implementation of org.freedesktop.impl.portal.desktop.wlr are all likely to be relevant.

It looks as though there is a 25 second pause after trying to activate xdg-desktop-portal-wlr by `systemd --user` activation. This is suspicious to me, because 25 seconds is the default timeout for a D-Bus method call. Probably something is going wrong there.

**Comment by smcv** [comment]
Also, please give version numbers: xdg-desktop-portal 1.17/1.18 has significant behaviour changes compared with earlier versions. ("The latest" is not a version number.)

**Comment by pescepalla** [comment]
```
xdg-desktop-portal 1.16.0-3
xdg-desktop-portal-wlr 0.7.0-1
gnome-keyring 1:42.1-3
```
And here's what happened in those 25 seconds:

```
Sep 27 13:45:02 stellaris systemd[1]: systemd-rfkill.service: Deactivated successfully.
Sep 27 13:45:07 stellaris syncthing[660]: [I4HKA] INFO: Established secure connection to B3HAY3X-6RGS3IQ-BQLXR54-RFOFEEH-OUIZZGK-4DFHOCU-KW5H77C-NNZ2GA6 at 192.168.1.102:50632-49.12.93.194:22067/relay-client/TLS1.3-TLS_AES_128_GCM_SHA256/WAN-P50
Sep 27 13:45:07 stellaris syncthing[660]: [I4HKA] INFO: Device B3HAY3X-6RGS3IQ-BQLXR54-RFOFEEH-OUIZZGK-4DFHOCU-KW5H77C-NNZ2GA6 client is "syncthing v1.23.7" named "FP4" at 192.168.1.102:50632-49.12.93.194:22067/relay-client/TLS1.3-TLS_AES_128_GCM_SHA256/WAN-P50
Sep 27 13:45:09 stellaris syncthing[660]: [I4HKA] INFO: quic://0.0.0.0:22000 detected NAT type: Symmetric NAT
Sep 27 13:45:11 stellaris dhcpcd[578]: wlan0: no IPv6 Routers available
Sep 27 13:45:20 stellaris syncthing[660]: [I4HKA] INFO: Detected 1 NAT service
Sep 27 13:45:25 stellaris dbus-daemon[884]: [session uid=971 pid=884] Activating via systemd: service name='org.gtk.vfs.Daemon' unit='gvfs-daemon.service' requested by ':1.0' (uid=971 pid=878 comm="regreet")
Sep 27 13:45:25 stellaris systemd[725]: Starting Virtual filesystem service...
Sep 27 13:45:25 stellaris dbus-daemon[884]: [session uid=971 pid=884] Activating service name='org.freedesktop.secrets' requested by ':1.2' (uid=971 pid=887 comm="/usr/lib/xdg-desktop-portal")
Sep 27 13:45:25 stellaris systemd[725]: Started GNOME Keyring daemon.
Sep 27 13:45:25 stellaris dbus-daemon[884]: [session uid=971 pid=884] Successfully activated service 'org.gtk.vfs.Daemon'
Sep 27 13:45:25 stellaris systemd[725]: Started Virtual filesystem service.
Sep 27 13:45:25 stellaris gnome-keyring-daemon[927]: GNOME_KEYRING_CONTROL=/run/user/971/keyring
Sep 27 13:45:25 stellaris gnome-keyring-daemon[927]: Gkm: unable to create keyring dir: /.local/share/keyrings
Sep 27 13:45:25 stellaris gnome-keyring-d[927]: unable to create keyring dir: /.local/share/keyrings
Sep 27 13:45:25 stellaris gnome-keyring-d[927]: unable to create keyring dir: /.local/share/keyrings
Sep 27 13:45:25 stellaris gnome-keyring-daemon[927]: Gkm: unable to create keyring dir: /.local/share/keyrings
Sep 27 13:45:25 stellaris dbus-daemon[884]: [session uid=971 pid=884] Successfully activated service 'org.freedesktop.secrets'
Sep 27 13:45:25 stellaris gnome-keyring-daemon[927]: The Secret Service was already initialized
Sep 27 13:45:25 stellaris gnome-keyring-d[927]: The Secret Service was already initialized
Sep 27 13:45:25 stellaris gnome-keyring-daemon[922]: discover_other_daemon: 1
Sep 27 13:45:25 stellaris org.freedesktop.secrets[922]: GNOME_KEYRING_CONTROL=/run/user/971/keyring
Sep 27 13:45:25 stellaris dbus-daemon[884]: [session uid=971 pid=884] Activating via systemd: service name='org.a11y.Bus' unit='at-spi-dbus-bus.service' requested by ':1.0' (uid=971 pid=878 comm="regreet")
Sep 27 13:45:25 stellaris systemd[725]: Starting Accessibility services bus...
Sep 27 13:45:25 stellaris dbus-daemon[884]: [session uid=971 pid=884] Successfully activated service 'org.a11y.Bus'
Sep 27 13:45:25 stellaris systemd[725]: Started Accessibility services bus.
```

**Comment by pescepalla** [comment]
I see this has been tagged with `needs more info`. Let me know if I can help.

**Comment by GeorgesStavracas** [comment]
Can you please test this with xdg-desktop-portal 1.18.0 or the `master` branch?

**Comment by GeorgesStavracas** [comment]
@pescepalla did you manage to test this with xdg-desktop-portal `main` branch, or at least version 1.18.2?

**Comment by pescepalla** [comment]
Hi, sorry for the late reply. Issue seems resolved as of 1.18.2.

**Comment by GeorgesStavracas** [comment]
Thanks for testing it :slightly_smiling_face:
